### PR TITLE
Add extended features to CGM adapter (#3002)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Added PR reference to merge commit title
- [ ] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by danielsotirhos** [comment]
Terminal log from a test run. 
[cgm_adapter_log.txt](https://github.com/DataBiosphere/azul/files/6386400/cgm_adapter_log.txt)

File: scripts/cgm_adapter.py [hunk 1]
```diff
@@ -279,30 +304,38 @@ def _run(self):
...
+        # If project's catalog was given on spreadsheet, check against that
+        project_catalog = project['catalog']
+        if project_catalog:
+            return project_catalog == catalog
+        # Otherwise ping Azul to check if the project exists in the catalog.
         else:
-            raise Exception(f'Unexpected response status {response.status_code!r} from {url!r}')
+            project_id = project['project_uuid']
+            url = furl('https://service.azul.data.humancellatlas.org/',
+                       path=f'/index/projects/{project_id}',
```
Comment (achave11) [inline scripts/cgm_adapter.py:330]:
```suggestion
                       path=('index', 'projects', project_id),
```
`furl.path` allows for usage of sequences, that way we don't have to use f-strings to concat.

File: scripts/cgm_adapter.py [hunk 2]
```diff
@@ -316,25 +349,43 @@ def parse_csv(self) -> Mapping[str, Any]:
...
+                require(file.new_name not in file_names,
+                        f'File {file.new_name!r} is not unique')
+                file_names.add(file.new_name)
+                if not uploaded == '1':
+                    log.info(f'Skipping row {row_num} file {file_name!r},'
+                             f' uploaded = {uploaded}')
+                    continue
+                if date_added and date_imported and date_imported >= date_added:
+                    log.info(f'Skipping row {row_num} file {file_name!r},'
+                             f' {date_imported!r} >= {date_added!r}')
```
Comment (achave11) [inline scripts/cgm_adapter.py:374]: Not sure if this logging statement can benefit from [CONTRIBUTING](https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#110logging) -> "We don't use f'' strings or string concatenation when interpolating dynamic values into log messages:"

**Review by achave11 — CHANGES_REQUESTED** [review]
Minor nits. It seems like the logging/log statements are a mix between f-string and %, I am curious if there is a reason for that.

**Comment by danielsotirhos** [comment]
> Minor nits. It seems like the logging/log statements are a mix between f-string and %, I am curious if there is a reason for that.

Sloppy code. I will covert the f-strings to `%` formatted strings.

**Review by achave11 — APPROVED** [review]
Looks good. You might want to also consider updating the format of the log messages in `scripts/sha256_gcs_blobs.py`, I think it's appropriate to address it in the same commit as s/logging/log, your call.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, pick any subset of features you'd like to demo.

**Pull request merged**